gnovm/pkg/gnolang/gonative.go

```diff
@@ -824,19 +824,25 @@ func gno2GoType(t Type) reflect.Type {
 			panic("should not happen")
 		}
 	case *PointerType:
-		et := gno2GoType(ct.Elem())
+		et := gno2GoType(ct.Elem(), seen)
 		return reflect.PointerTo(et)
 	case *ArrayType:
 		ne := ct.Len
-		et := gno2GoType(ct.Elem())
+		et := gno2GoType(ct.Elem(), seen)
 		return reflect.ArrayOf(ne, et)
 	case *SliceType:
-		et := gno2GoType(ct.Elem())
+		et := gno2GoType(ct.Elem(), seen)
 		return reflect.SliceOf(et)
 	case *StructType:
+		if val, ok := seen[t]; ok {
+			return val
+		}
+
+		placeholder := reflect.TypeOf(struct{}{})
+		seen[t] = placeholder
```

Comment:
Actually, nevermind.

I just realised this just replaces structs with `struct{}`. It creates a panic whenever the recursive structure is non-nil.

And, because of maps, this doesn't work:

```go
package main

import "fmt"

type Normal struct {
	A Role
	B Role
}

type Role struct {
	Name        string
	Permissions []string
	Users       []string
	Next        *Role
	Prev        *Role
}

func main() {
	userRole := &Role{
		Name:        "user",
		Permissions: []string{},
		Users:       []string{},
		Next:        nil,
		Prev:        nil,
	}

	n := Normal{A: Role{Name: "a"}, B: Role{Name: "b"}}
	_ = userRole

	fmt.Printf("%v", n)
}

// Output:
// &{user [] [] <nil> <nil>}
```

Proper resolution of this requires named type support in reflect, and that's not possible. Seeing the plan to remove gonative, my advice is to drop this issue entirely as a "won't fix".

Apologies, I know I mentioned it was a good effort, but I think I may have misread/misunderstood when I read the original issue.

Reply:
No worries, I understand, thanks for the feedback. :heart: